compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -319,7 +319,37 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     }
                 }
                 ItemKind::TyAlias(self_ty, _) => icx.to_ty(self_ty),
-                ItemKind::Impl(hir::Impl { self_ty, .. }) => icx.to_ty(*self_ty),
+                ItemKind::Impl(
+                    hir::Impl { self_ty, .. }
+                ) => {
+                    struct MyVisitor(bool);
+                    impl<'v> hir::intravisit::Visitor<'v> for MyVisitor {
+                        fn visit_ty(&mut self, t: &'v Ty<'v>) {
+                            if matches!(
+                                &t.kind,
+                                TyKind::Path(hir::QPath::Resolved(
+                                    _,
+                                    Path {
+                                        res: hir::def::Res::SelfTyAlias { .. },
+                                        ..
+                                    },
+                                ))
+                            ) {
+                                self.0 = true;
+                                return;
+                            }
+                            hir::intravisit::walk_ty(self, t);
+                        }
+                    }
+
+                    let mut my_visitor = MyVisitor(false);
+                    my_visitor.visit_ty(self_ty);
```

Comment:
seconded Esteban's idea about making this return a `Span` -- could you maybe pull this logic into a separate function named something like `find_self_alias_in_ty` with `-> Option<Span>`, and then we can match on that to deny the self alias?